src/target/riscv/riscv-013.c

```diff
@@ -279,16 +281,21 @@ dm013_info_t *get_dm(struct target *target)
 	return dm;
 }
 
-static uint32_t set_hartsel(uint32_t initial, uint32_t index)
+static uint32_t set_hartsel(uint32_t initial, int index)
 {
+	initial &= ~DM_DMCONTROL_HASEL;
 	initial &= ~DM_DMCONTROL_HARTSELLO;
 	initial &= ~DM_DMCONTROL_HARTSELHI;
 
-	uint32_t index_lo = index & ((1 << DM_DMCONTROL_HARTSELLO_LENGTH) - 1);
-	initial |= index_lo << DM_DMCONTROL_HARTSELLO_OFFSET;
-	uint32_t index_hi = index >> DM_DMCONTROL_HARTSELLO_LENGTH;
-	assert(index_hi < 1 << DM_DMCONTROL_HARTSELHI_LENGTH);
-	initial |= index_hi << DM_DMCONTROL_HARTSELHI_OFFSET;
+	if (index > 0) {
+		uint32_t index_lo = index & ((1 << DM_DMCONTROL_HARTSELLO_LENGTH) - 1);
+		initial |= index_lo << DM_DMCONTROL_HARTSELLO_OFFSET;
+		uint32_t index_hi = index >> DM_DMCONTROL_HARTSELLO_LENGTH;
+		assert(index_hi < 1 << DM_DMCONTROL_HARTSELHI_LENGTH);
+		initial |= index_hi << DM_DMCONTROL_HARTSELHI_OFFSET;
+	} else if (index < 0) {
+		initial |= DM_DMCONTROL_HASEL;
```

Comment:
Caution: If `index==-1`, hart #0 will always get selected - because in that case `dmcontrol.hartsel` will contain value `0`. 

Will this work as expected in all cases? What if we have SMP+hwthread configuration on harts 1,3,5, for instance, and hart 0 is not part of the SMP group?

Reply:
You're right. This will not always work. This bug would have existed before this commit as well. I don't want to fix this right now, but I filed #748 to track it.